### PR TITLE
Support k8s-rdma-shared-device-plugin device selectors

### LINKED ISSUE #35: [Enhancement] Support k8s-rdma-shared-device-plugin device selectors
Once a release is made for k8s-rdma-shared-device-plugin
we can leverage device selectors feature[1] to simplify deployment (by not relying on the netdev to create an RDMA resource),
we should support device selectors configuration in helm chart and update README.

[1] https://github.com/Mellanox/k8s-rdma-shared-dev-plugin/pull/19/commits

### PR DESCRIPTION
**Pull request by Mmduh-483**
Fixes #35

### CODE REVIEW

**Comment by Mmduh-483** [comment]
@adrianchiris @moshe010 Still needs to update the shared dev plugin release

File: deployment/network-operator/values.yaml [hunk 1]
```diff
@@ -84,4 +84,6 @@ devicePlugin:
   # example:
   resources:
     - name: rdma_shared_device_a
-      devices: [enp5s0f0]
\ No newline at end of file
+      vendors: [15b3]
+      deviceIDs: ["1017"]
+      ifNames: [enp5s0f0]
```
Comment (adrianchiris) [inline deployment/network-operator/values.yaml:89]: missing newline

File: example/deploy/crds/mellanox.com_v1alpha1_nicclusterpolicy_cr.yaml [hunk 2]
```diff
@@ -31,8 +31,12 @@ spec:
         "configList": [
           {
             "resourceName": "hca_shared_devices_a",
-            "rdmaHcaMax": 1000,
-            "devices": ["ens2f0"]
+            "rdmaHcaMax": 100,
+            "selectors": {
+              "vendors": ["15b3"],
```
Comment (adrianchiris) [inline example/deploy/crds/mellanox.com_v1alpha1_nicclusterpolicy_cr.yaml:36]: can you keep just ifNames in the example ?

File: deployment/network-operator/README.md [hunk 3]
```diff
@@ -165,7 +169,7 @@ devicePlugin:
   deploy: true
   reources:
     - name: rdma_shared_device_a
-      devices: [enp1, enp2]
+      ifNames: [enp1, enp2]
```
Comment (adrianchiris) [inline deployment/network-operator/README.md:172]: can you add an example that with vendor and deviceID selectors that maps mellanox cx5 under a resource

File: deployment/network-operator/README.md [hunk 4]
```diff
@@ -114,9 +114,13 @@ to be associated with the resource.
 ```
 resources:
     - name: rdma_shared_device_a
-      devices: [enp5s0f0]
+      vendors: [15b3]
```
Comment (adrianchiris) [inline deployment/network-operator/README.md:117]: the description of this section should be updated

File: deployment/network-operator/values.yaml (hunk 1 above)
Comment (adrianchiris) [reply deployment/network-operator/values.yaml:89]: lets change the default to something that will be usable out of the box.
```
name: rdma_shared_device
vendors: [15b3]
```

File: deployment/network-operator/templates/mellanox.com_v1alpha1_nicclusterpolicy_cr.yaml [hunk 5]
```diff
@@ -48,7 +48,11 @@ spec:
           {
             "resourceName": {{ $element.name | quote }},
             "rdmaHcaMax": 1000,
-            "devices": {{ $element.devices | toJson }}
+            "selectors": {
+              "vendors": {{ $element.vendors | toJson }},
```
Comment (adrianchiris) [inline deployment/network-operator/templates/mellanox.com_v1alpha1_nicclusterpolicy_cr.yaml:52]: what happens if some attributes are missing  ? will it be empty ?

**Review by adrianchiris — CHANGES_REQUESTED** [review]

File: deployment/network-operator/values.yaml (hunk 1 above)
Comment (Mmduh-483) [reply deployment/network-operator/values.yaml:89]: changed

File: example/deploy/crds/mellanox.com_v1alpha1_nicclusterpolicy_cr.yaml (hunk 2 above)
Comment (Mmduh-483) [reply example/deploy/crds/mellanox.com_v1alpha1_nicclusterpolicy_cr.yaml:36]: changed

File: deployment/network-operator/README.md (hunk 3 above)
Comment (Mmduh-483) [reply deployment/network-operator/README.md:172]: added

File: deployment/network-operator/README.md (hunk 4 above)
Comment (Mmduh-483) [reply deployment/network-operator/README.md:117]: Updated

File: deployment/network-operator/templates/mellanox.com_v1alpha1_nicclusterpolicy_cr.yaml (hunk 5 above)
Comment (Mmduh-483) [reply deployment/network-operator/templates/mellanox.com_v1alpha1_nicclusterpolicy_cr.yaml:52]: Updated to be empty list

File: deployment/network-operator/README.md [hunk 6]
```diff
@@ -108,15 +108,19 @@ We have introduced the following Chart parameters.
 
 ##### RDMA Device Plugin Resource configurations
 
-Consists of a list of RDMA resources each with a name a a list of RDMA capable network devices
-to be associated with the resource.
+Consists of a list of RDMA resources each with a name and selector of RDMA capable network devices
+to be associated with the resource, refer to [RDMA Shared Device Plugin Selectors](https://github.com/Mellanox/k8s-rdma-shared-dev-plugin#devices-selectors) for supported selectors.
```
Comment (adrianchiris) [inline deployment/network-operator/README.md:112]: nit: ", refer" -> ". Refer"

File: deployment/network-operator/README.md (hunk 3 above)
Comment (adrianchiris) [reply deployment/network-operator/README.md:172]: I meant add an example -> new example that demonstrates it.

you can have this example demonstrate a network operator deployment of device plugin only that creates a single resource which maps to mellanox cx5

File: deployment/network-operator/templates/mellanox.com_v1alpha1_nicclusterpolicy_cr.yaml (hunk 5 above)
Comment (Mmduh-483) [reply deployment/network-operator/templates/mellanox.com_v1alpha1_nicclusterpolicy_cr.yaml]: updated

File: deployment/network-operator/README.md [hunk 7]
```diff
@@ -165,7 +169,26 @@ devicePlugin:
...
-      devices: [enp1, enp2]
+      ifNames: [enp1, enp2]
     - name: rdma_shared_device_b
-      devices: [ib0]
+      ifNames: [ib0]
+```
+
+#### Example 3
+Network Operator deployment with the default version of OFED and NV Peer mem driver, RDMA device
+plugin with two RDMA resources, mapped to Mellanox ConnectX-5.
```
Comment (adrianchiris) [inline deployment/network-operator/README.md:179]: The description does not match the yaml below

also can we have example 3 to deploy only device plugin ?


from my comment above
> you can have this example demonstrate a network operator deployment of device plugin only

**Review by adrianchiris — CHANGES_REQUESTED** [review]
@Mmduh-483 please adjust example 3 and we can merge

**Review by adrianchiris — CHANGES_REQUESTED** [review]
my last comment was not addressed

**Comment by adrianchiris** [comment]
@Mmduh-483 i have created a release of k8s-rdma-shared-dp which contains support for device selectors.
Can you rebase and update the PR accordingly ?

**Comment by adrianchiris** [comment]
@Mmduh-483 need also to update the deployment version (tag) of device plugin to be `v1.1.0`

File: deployment/network-operator/README.md [hunk 8]
```diff
@@ -204,9 +208,24 @@ devicePlugin:
...
   reources:
     - name: rdma_shared_device_a
-      devices: [enp1, enp2]
+      ifNames: [enp1, enp2]
     - name: rdma_shared_device_b
-      devices: [ib0]
+      ifNames: [ib0]
+```
+
+#### Example 3
```
Comment (adrianchiris) [inline deployment/network-operator/README.md:216]: you now have two Example 3 in the doc, also example 3 below is outdated

**Review by adrianchiris — CHANGES_REQUESTED** [review]

File: deployment/network-operator/README.md (hunk 8 above)
Comment (Mmduh-483) [reply deployment/network-operator/README.md:216]: fixed

**Review by adrianchiris — APPROVED** [review]

**Pull request merged**